lib/improver/nbhood/recursive_filter.py

```diff
@@ -0,0 +1,394 @@
+                backwards x-direction.
+        """
+
+        x_lim = grid.shape[0]
+        for i in range(x_lim-2, -1, -1):
+            grid[i, :] = ((1. - alphas[i, :]) * grid[i, :] +
+                          alphas[i, :] * grid[i+1, :])
+        return grid
+
+    @staticmethod
+    def recurse_forward_y(grid, alphas):
+        """
+        Method to run the recursive filter in the forward y-direction.
+
+        In the forward direction:
+            Recursive filtering is calculated as:
+                Bi = ((1-alpha) * Ai) + (alpha * Bi-1)
+
+            Progressing from gridpoint i-1 to i:
+                Bi = new value at gridpoint i, Ai = Old value at gridpoint i
+                Bi-1 = New value at gridpoint i-1
+
+        Args:
+            grid (numpy array):
+                Array containing the input data to which the recursive filter
+                will be applied.
+            alphas (numpy array):
+                Array of alpha values that will be used when applying
+                the recursive filter along the y-axis.
+
+        Returns:
+            grid (numpy array):
+                Array containing the smoothed field after the recursive
+                filter method has been applied to the input array in the
+                forward y-direction.
+        """
+
+        y_lim = grid.shape[1]
+        for i in range(1, y_lim):
+            grid[:, i] = ((1. - alphas[:, i]) * grid[:, i] +
+                          alphas[:, i] * grid[:, i-1])
+        return grid
+
+    @staticmethod
+    def recurse_backwards_y(grid, alphas):
+        """
+        Method to run the recursive filter in the backwards y-direction.
+
+        In the backwards direction:
+            Recursive filtering is calculated as:
+                Bi = ((1-alpha) * Ai) + (alpha * Bi+1)
+
+            Progressing from gridpoint i+1 to i:.
+                Bi = new value at gridpoint i, Ai = Old value at gridpoint i
+                Bi+1 = New value at gridpoint i+1
+
+        Args:
+            grid (numpy array):
+                Array containing the input data to which the recursive filter
+                will be applied.
+            alphas (numpy array):
+                Array of alpha values that will be used when applying
+                the recursive filter along the y-axis.
+
+        Returns:
+            grid (numpy array):
+                Array containing the smoothed field after the recursive
+                filter method has been applied to the input array in the
+                backwards y-direction.
+        """
+
+        y_lim = grid.shape[1]
+        for i in range(y_lim-2, -1, -1):
+            grid[:, i] = ((1. - alphas[:, i]) * grid[:, i] +
+                          alphas[:, i] * grid[:, i+1])
+        return grid
+
+    @staticmethod
+    def run_recursion(cube, alphas_x, alphas_y, iterations):
+        """
+        Method to run the recursive filter.
```

Comment:
I don't think we need the same method information in both the run_recursion function and the individual functions. Just leave it in the individual functions now you've added it there and remove it from here.

Reply:
Agreed.